### ISSUE TITLE
Unexpected printer status code (Kyocera FS-C5100DN)

### ISSUE THREAD

**Issue by tschamm** [ctalkington/python-ipp#14]
```
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/data_entry_flow.py", line 109, in async_init
    result = await self._async_handle_step(flow, flow.init_step, data)
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/data_entry_flow.py", line 172, in _async_handle_step
    result: Dict = await getattr(flow, method)(user_input)
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/components/ipp/config_flow.py", line 110, in async_step_zeroconf
    info = await validate_input(self.hass, self.discovery_info)
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/components/ipp/config_flow.py", line 46, in validate_input
    printer = await ipp.printer()
  File "/srv/homeassistant/lib/python3.7/site-packages/pyipp/ipp.py", line 211, in printer
    "requested-attributes": DEFAULT_PRINTER_ATTRIBUTES,
  File "/srv/homeassistant/lib/python3.7/site-packages/pyipp/ipp.py", line 189, in execute
    {"status-code": parsed["status-code"]},
pyipp.exceptions.IPPError: ('Unexpected printer status code', {'status-code': 1283})
```

**Comment by ctalkington** [comment]
Thanks for the report. Do you happen to have the ability to run python script to get a dump of the data provided by printer?

**Comment by ctalkington** [comment]
It appeara that your printer may have enhanced security as looking at a manual it says:

If you are using NetWare, AppleTalk, IPP, SSL Server or IPSec to connect to a network, you must enter an administrator password.  By default, the administrator password is the numeric part of your printer's model name. If you use FS-C5300DN, enter "5300"

If this is thecase and its not something else, I will have to look into supporting passwords if this is the case and discovery would need to try and detect the auth needed.

**Comment by tschamm** [comment]
Thanks for the hints. Here is the output from the `examples/printer.py` script:
```
Traceback (most recent call last):
  File "examples/printer.py", line 23, in <module>
    loop.run_until_complete(main())
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "examples/printer.py", line 12, in main
    printer = await ipp.printer()
  File "/Users/schamm/Development/python-ipp/pyipp/ipp.py", line 208, in printer
    "requested-attributes": DEFAULT_PRINTER_ATTRIBUTES,
  File "/Users/schamm/Development/python-ipp/pyipp/ipp.py", line 176, in execute
    response = await self._request(data=message)
  File "/Users/schamm/Development/python-ipp/pyipp/ipp.py", line 141, in _request
    "status-code": response.status,
pyipp.exceptions.IPPResponseError: ('HTTP 404', {'content-type': 'text/html', 'message': ' 404 Page Not Found', 'status-code': 404})
```
I'm still struggeling getting a valid connection to the printer via `python-ipp`. Connecting to the printer via the OS (Mac OS), at least I don't need to enter a password.

**Comment by tschamm** [comment]
I assume the unexpected printer status `pyipp.exceptions.IPPError: ('Unexpected printer status code', {'status-code': 1283})` happens from sharing the printers on my computer via network (first result of the zeroconf discovery run).

Here's the result from a zeroconf discovery run of `"_ipp._tcp.local."`:
```
Service Kyocera FS-C5100DN @ Cooper._ipp._tcp.local. added, service info: ServiceInfo(type='_ipp._tcp.local.', name='Kyocera FS-C5100DN @ Cooper._ipp._tcp.local.', addresses=[b'\xc0\xa8\x01('], port=631, weight=0, priority=0, server='Cooper-319.local.', properties={b'txtvers': b'1', b'qtotal': b'1', b'rp': b'printers/KM448FAC', b'ty': b'Kyocera Generic Color A4 PCL', b'adminurl': b'https://Cooper-319.local.:631/printers/KM448FAC', b'priority': b'0', b'product': b'(FS-C5100DN)', b'pdl': b'application/octet-stream,application/pdf,application/postscript,image/jpeg,image/png,image/pwg-raster', b'UUID': b'2c609d50-6399-3b68-72d5-f2c98ecd777d', b'TLS': b'1.2', b'Color': b'T', b'Duplex': b'T', b'Copies': b'T', b'Collate': b'T', b'printer-state': b'3', b'printer-type': b'0x8090DE'})
Service KM448FAC._ipp._tcp.local. added, service info: ServiceInfo(type='_ipp._tcp.local.', name='KM448FAC._ipp._tcp.local.', addresses=[b'\xc0\xa8\x01\x05'], port=631, weight=0, priority=0, server='KM448FAC.local.', properties={b'pdl': b'application/postscript', b'product': b'(FS-C5100DN)', b'note': b'', b'ty': b'Kyocera FS-C5100DN', b'txtvers': b'1', b'priority': b'48', b'rp': b'printers/lp1', b'qtotal': b'1', b'usb_MFG': b'Kyocera', b'usb_MDL': b'Kyocera FS-C5100DN (KPDL)', b'adminurl': b'http://192.168.1.5'})
```

**Comment by ctalkington** [comment]
Did you happen to change the address in the example python? From what im seeing you would try:

ipp://KM448FAC.local:631/printers/lp1 for the printer itself

ipps://Cooper-319.local:631/printers/KM448FAC for the mac shared

There is also a debug example that generates a dump "bin" file if you get either of those to respond in the example script. I can use the bin file to test against if provided

**Comment by tschamm** [comment]
For the mac shared printer, using the uri `ipps://Cooper-319.local:631/printers/KM448FAC` works. The shared printer is also listed successful in home assistant.

**Comment by ctalkington** [comment]
Out of curiousity did any printers get offered via discovery?

**Comment by tschamm** [comment]
Yes, the mac shared printer was offered via discovery. 

For the printer itself, I get multiple errors.
```
2020-04-13 23:50:27 ERROR (MainThread) [homeassistant.components.ipp.config_flow] IPP Parse Error
Traceback (most recent call last):
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/ipp.py", line 219, in printer
    printer = Printer.from_dict(parsed)
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/models.py", line 153, in from_dict
    for marker_id in range(len(marker_names))
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/models.py", line 153, in <listcomp>
    for marker_id in range(len(marker_names))
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/components/ipp/config_flow.py", line 110, in async_step_zeroconf
    info = await validate_input(self.hass, self.discovery_info)
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/components/ipp/config_flow.py", line 46, in validate_input
    printer = await ipp.printer()
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/ipp.py", line 221, in printer
    raise IPPParseError from exc
pyipp.exceptions.IPPParseError
2020-04-13 23:50:27 ERROR (MainThread) [homeassistant.components.ipp.config_flow] IPP Parse Error
Traceback (most recent call last):
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/ipp.py", line 219, in printer
    printer = Printer.from_dict(parsed)
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/models.py", line 153, in from_dict
    for marker_id in range(len(marker_names))
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/models.py", line 153, in <listcomp>
    for marker_id in range(len(marker_names))
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/components/ipp/config_flow.py", line 110, in async_step_zeroconf
    info = await validate_input(self.hass, self.discovery_info)
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/components/ipp/config_flow.py", line 46, in validate_input
    printer = await ipp.printer()
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/ipp.py", line 221, in printer
    raise IPPParseError from exc
pyipp.exceptions.IPPParseError
```
and later the one I thought is relevant
```
2020-04-13 23:50:37 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/data_entry_flow.py", line 106, in async_init
    result = await self._async_handle_step(flow, flow.init_step, data)
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/data_entry_flow.py", line 169, in _async_handle_step
    result: Dict = await getattr(flow, method)(user_input)
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/components/ipp/config_flow.py", line 110, in async_step_zeroconf
    info = await validate_input(self.hass, self.discovery_info)
  File "/Users/schamm/Development/home-assistant/ha-core/homeassistant/components/ipp/config_flow.py", line 46, in validate_input
    printer = await ipp.printer()
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/ipp.py", line 211, in printer
    "requested-attributes": DEFAULT_PRINTER_ATTRIBUTES,
  File "/Users/schamm/Development/home-assistant/ha-core/venv/lib/python3.7/site-packages/pyipp/ipp.py", line 189, in execute
    {"status-code": parsed["status-code"]},
pyipp.exceptions.IPPError: ('Unexpected printer status code', {'status-code': 1283})
```

**Comment by ctalkington** [comment]
Whst verdion of pyipp, pretty sure the marker error shouldnt be happening

**Comment by tschamm** [comment]
HA: Home Assistant 0.109.0.dev0 (Mac) and 0.108.3 (Raspberry).
pyipp: pyipp-0.9.1 (Mac) and pyipp-0.9.2 (Raspberry).

**Comment by ctalkington** [comment]
hum yah that would of been pyipp 0.9.2 that fix was in so that might explain issue on mac with markers. However the printer bin file you sent was basically empty so wonder if thats part of the issue

**Comment by tschamm** [comment]
Only the mac shared printer shows up.

**Comment by ctalkington** [comment]
Ok so the pronter itself isn't secure itself, just the mac provided printer. Let me see if i can tell what the status code means

**Comment by ctalkington** [comment]
ah ok this makes more sense.
So 1283 is 0x0503 which is:

server-error-version-not-supported (0x0503)
The IPP object does not support or refuses to support the IPP version that was supplied as the value of the "version-number" operation parameter in the request.  The IPP object is indicating that it is unable or unwilling to complete the request using the same major and minor version number as supplied in the request,

In other words your printer isnt capable of IPP 2.0 which is what this python module is using as a basis as it offers the most relevant data whereas IPP 1.1 is a bit more limited. In this case its best to use the mac shared printer since its essentially serving an IPP 2.0 compatible response.

**Comment by ctalkington** [comment]
Im going to see about catching that error code and raising a specific error so that we can abort with a better message in HA

**Comment by tschamm** [comment]
That's fine for me.
Thanks a lot for your support! I think we can close this issue then.
BTW: upgrading to 0.9.2 helped to get rid of the markers error.
